src/rules/unnecessary_function_arguments.erl

```diff
@@ -3,24 +3,88 @@
 %%      ignored in all function clauses.</p>
 %%      <p>To avoid this warning, remove the unused argument(s).</p>
 %%      <p><b>Note:</b> This rule will not emit a warning if the function
-%%      implements a behaviour callback or a NIF call.</p>
+%%      implements a NIF call or local or dynamic behaviour callback.
+%%      That said, for the majority of the OTP behaviours implementations this rule
+%%      will be applied.</p>
 -module(unnecessary_function_arguments).
 
 -behaviour(hank_rule).
 
 -export([analyze/2, ignored/2]).
 
+%% Known OTP behaviours which do not implement dynamic callbacks like ct_suite.
+-define(KNOWN_BEHAVIOURS,
+        [application,
+         gen_event,
+         gen_server,
+         ssh_channel,
+         ssh_client_channel,
+         ssh_client_key_api,
+         ssh_server_channel,
+         ssh_server_key_api,
+         ssl_crl_cache_api,
+         ssl_session_cache_api,
+         supervisor,
+         supervisor_bridge,
+         tftp]).
+
+%% Allow erl_syntax:syntaxTree/0 type spec
+%% Allow Module:behaviour_info/1 call
+-elvis([{elvis_style, invalid_dynamic_call, disable},
+        {elvis_style, nesting_level, [5]},
+        {elvis_style, atom_naming_convention, #{regex => "^([a-zA-Z][a-z0-9]*_?)*$"}}]).
+
+-type imp_callbacks() :: #{File :: string() => [tuple()] | ignore}.
+
 %% @private
 -spec analyze(hank_rule:asts(), hank_context:t()) -> [hank_rule:result()].
 analyze(FilesAndASTs, _Context) ->
+    ImpCallbacks = callback_usage(FilesAndASTs),
```

Comment:
What if instead of collecting all the callback implementation information beforehand we do it per-file within the list comprehension (i.e. in `is_callback/3`)?
Not sure if that would help reduce the complexity of the code or not… just a random thought.

Reply:
I think that's not worth it here, since it's using `not is_unrecognized_behaviour/2` to detect if ignoring the whole file, and `is_callback` is used to analyze a Node/Function inside the File.